scheduler/src/cook/rate_limit/generic.clj

```diff
@@ -0,0 +1,108 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.rate-limit.generic
+  (:require [clj-time.core]
+            [clj-time.coerce]
+            [cook.rate-limit.token-bucket-filter :as tbf])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (spend! [this key resources]
+    "Request this number of resources. Does not block (or otherwise indicate we're in debt)")
+  (time-until-out-of-debt-millis! [this key]
+    "Time, in milliseconds, until this rate limiter is out of debt for this key. Earns resources first.")
+  (enforce? [this]
+    "Are we enforcing this policy?"))
+
+(defn current-time-in-millis
+  []
+  (clj-time.coerce/to-long (clj-time.core/now)))
+
+(defn get-key
+  "Guava cache key cannot be nil/null"
+  [key]
+  (or key "*NULL_TBF_KEY*"))
+
+(defn get-token-bucket-filter
+  "Get or create a TBF for the requested key. Buckets start off full, so users whose first request is a lot of
+  fast requests don't get throttled immediately. This doesn't let users get much above their quota ---
+  The fact that we don't have a TBF says that we are either just starting up (not likely) or haven't
+  gotten anything from them in a long time. (Assuming the TTL is set at least as high as
+  (:bucket-size/:tokens-replenished-per-minute))."
+  [limiter key]
+  (let [{:keys [tokens-replenished-per-minute max-tokens]} limiter
+        supplier (proxy [Callable] []
```

Comment:
Why use `proxy` here instead of `reify`?

Reply:
No reason. Thanks! Fixed.